bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/SpecificPageFilterFactory.java

```diff
@@ -0,0 +1,103 @@
+ *
+ ******************************************************************************/
+
+package com.adobe.cq.commerce.core.components.internal.servlets;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.metatype.annotations.Designate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.day.cq.commons.jcr.JcrConstants;
+import com.day.cq.wcm.api.NameConstants;
+
+@Component(
+    property = {
+        "sling.filter.scope=REQUEST",
+        "sling.filter.methods=GET",
+        "sling.filter.extensions=html"
+    })
+@Designate(ocd = SpecificPageFilterConfiguration.class, factory = true)
+public class SpecificPageFilterFactory implements Filter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SpecificPageFilterFactory.class);
+
+    private static final String SELECTOR_FILTER_PROPERTY = "selectorFilter";
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {}
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        SlingHttpServletRequest slingRequest = (SlingHttpServletRequest) request;
+
+        // Skip filter if there isn't any selector in the URL
+        String selector = slingRequest.getRequestPathInfo().getSelectorString();
+        if (selector == null) {
+            chain.doFilter(request, response);
+            return;
+        }
+
+        Resource page = slingRequest.getResource();
+        LOGGER.debug("Checking sub-pages for {}", slingRequest.getRequestURI());
+
+        Iterator<Resource> children = page.listChildren();
+        while (children.hasNext()) {
+            Resource child = children.next();
+            if (!NameConstants.NT_PAGE.equals(child.getResourceType())) {
+                continue;
+            }
+
+            Resource jcrContent = child.getChild(JcrConstants.JCR_CONTENT);
+            if (jcrContent == null) {
+                continue;
+            }
+
+            Object filter = jcrContent.getValueMap().get(SELECTOR_FILTER_PROPERTY);
+            if (filter == null) {
+                continue;
+            }
+
+            // The property is saved as a String when it's a simple selection, or an array when a multi-selection is done
+            String[] selectors = filter.getClass().isArray() ? ((String[]) filter) : ArrayUtils.toArray((String) filter);
+
+            if (ArrayUtils.contains(selectors, selector)) {
```

Comment:
The logic will match and pick always the first child page which matches the selector. In case an author selects one product/category on multiple template pages always the first will be picked. This has to be documented as well.

Reply:
FYI, picking the first matching child was defined in the JIRA issue, but yes, I will add some documentation
> First generic category and product pages are default